### DISCUSSION TITLE
custom parser for multiple keys

### DISCUSSION THREAD

**Discussion by tonyxiao** [47ng/nuqs#463] [Ideas] [answered]
It's possible to implement a custom parser for a single key, but sometimes it is super handy to have a parser that transforms values from multiple keys together into the desired format.

**Comment by franky47** [comment, marked as answer]
Do you have an example?

You can kind of get there with `useQueryStates`, and declaring an object of parsers for the individual keys:

```tsx
const searchParamsParsers = {
  page: parseAsInteger.withDefault(1),
  modalOpen: parseAsBoolean.withDefault(false),
  search: parseAsString
}

const [{ page, modalOpen, search }, updateSearchParams] = useQueryStates(searchParamsParsers)

const nextPage = () => updateSearchParams(state => ({
  page: state.page + 1
  // Partial update: other search params keep their values
}))
```

One downside to this is that every key is parsed individually. It would not help in cases where you need key `foo` to have a particular set of values when `bar` has a specific value. Adding a level of validation on the resulting object could be a good feedback to add to #446.
